### PR TITLE
Add search command

### PR DESCRIPTION
**Pull request by mrmacete**
- agent can run r2 commands on running main RCore instance
- search parameters (from, to, flags, prefix, count, kwidx) are honored
- search results are set as flags (if enabled by config)
- add inner "search.in" config variable, with its help

example session:

```
$ r2 frida://Telegram
[0x00000000]> \e search.in=path:Telegram.app

[0x00000000]> \/ Telegram FAQ
Searching 12 bytes: 54 65 6c 65 67 72 61 6d 20 46 41 51
Searching 12 bytes in [0x0000000103b7a000-0x0000000106cdb000]
Searching 12 bytes in [0x0000000106cdb000-0x0000000106e53000]
Searching 12 bytes in [0x0000000107177000-0x000000010727c000]
hits: 4
0x105bb0b73 hit0_0 Telegram FAQ? Dort findest du
0x105bb0c31 hit0_1 Telegram FAQ";"lng_settings_l
0x105bd689f hit0_2 Telegram FAQ{faq_close} 에서
0x105bf3317 hit0_3 Telegram FAQ{faq_close} voor m
[0x00000000]> s 0x105bb0b73
[0x105bb0b73]> \e search.in=current

[0x105bb0b73]> \/ http
Searching 4 bytes: 68 74 74 70
Searching 4 bytes in [0x0000000103b7a000-0x0000000106cdb000]
hits: 182
[...]
0x1059ffce0 hit1_12 http://www.opengis.net/gml/3.2
0x105a23746 hit1_13 http://www.abisource.com/awml.
0x105a43d2e hit1_14 http://www.gribuser.ru/xml/fic
0x105a448ab hit1_15 http://www.lysator.liu.se/~all
0x105a450d4 hit1_16 http://www.daa.com.au/~james/d
[...]
[0x105bb0b73]>
```

should close https://github.com/nowsecure/r2frida/issues/43

### CODE REVIEW

**Comment by mrmacete** [comment]
To make this work, i added compatibility for commands which return a promise, if anyone is happy with that, could be used in java class list too and avoid the sync wait ?

**Comment by radare** [comment]
Sounds good! But maybe in another pr after merging this.

Thanks! I'll review now

File: Makefile [hunk 1]
```diff
@@ -22,6 +22,8 @@ LDFLAGS+=-shared -fPIC
 # R2
 CFLAGS+=$(shell pkg-config --cflags r_io)
 LDFLAGS+=$(shell pkg-config --libs r_io)
+CFLAGS+=$(shell pkg-config --cflags r_core)
```
Comment (radare) [inline Makefile]: r_core includes r_io already, so there's no need to include r_io, remove the lines above

File: src/agent/index.js [hunk 2]
```diff
@@ -307,18 +312,76 @@ const __environ = Memory.readPointer(Module.findExportByName(null, 'environ'));
...
+const configHelp = {
+  'search.in': configHelpSearchIn
+};
+
+const configValidator = {
+  'search.in': configValidateSearchIn
+};
+
+function configHelpSearchIn () {
+  return [
```
Comment (radare) [inline src/agent/index.js]: use backtick quote instead of array+join

File: src/agent/index.js [hunk 3]
```diff
@@ -307,18 +312,76 @@ const __environ = Memory.readPointer(Module.findExportByName(null, 'environ'));
...
+    'current         search the range containing current offset',
+    'path:pattern    search ranges mapping paths containing \'pattern\'',
+    ''
+  ].join('\n');
+}
+
+function configValidateSearchIn (val) {
+  const valSplit = val.split(':');
+  const [scope, param] = valSplit;
+  if (param !== undefined) {
```
Comment (radare) [inline src/agent/index.js]: use positive conditionals. `if (param === undefined) { ...`

File: src/agent/index.js [hunk 4]
```diff
@@ -307,18 +312,76 @@ const __environ = Memory.readPointer(Module.findExportByName(null, 'environ'));
...
+    if (scope === 'perm') {
+      const paramSplit = param.split('');
+      if (paramSplit.length !== 3 || valSplit.length > 2) {
+        return false;
+      }
+      const [r,w,x] = paramSplit;
```
Comment (radare) [inline src/agent/index.js]: spaces after commas

File: src/agent/index.js (hunk 3 above)
Comment (radare) [inline src/agent/index.js]: you can return early and avoid the else block to reduce one level of indentation.

File: src/agent/index.js [hunk 5]
```diff
@@ -307,18 +312,76 @@ const __environ = Memory.readPointer(Module.findExportByName(null, 'environ'));
...
+      return (r === 'r' || r === '-') &&
+        (w === 'w' || w === '-') &&
+        (x === 'x' || x === '-');
+    }
+    if (scope === 'path') {
```
Comment (radare) [inline src/agent/index.js]: return (scope === 'path');

File: src/agent/index.js (hunk 4 above)
Comment (mrmacete) [inline src/agent/index.js]: right. do we want to bring semistandard in?

File: src/agent/index.js [hunk 6]
```diff
@@ -307,18 +312,76 @@ const __environ = Memory.readPointer(Module.findExportByName(null, 'environ'));
...
+    .map(k => 'e ' + k + '=' + config[k])
     .join('\n');
   }
   const kv = args[0].split(/=/);
   if (kv.length === 2) {
     if (config[kv[0]] !== undefined) {
+      if (kv[1] === '?') {
+        if (configHelp[kv[0]] !== undefined) {
+          return configHelp[kv[0]]();
+        } else {
```
Comment (radare) [inline src/agent/index.js]: no need for an else block if you return.

File: src/agent/index.js [hunk 7]
```diff
@@ -1313,6 +1383,217 @@ function fridaVersion() {
...
+      hit.content = _byteArrayToHex(bytes);
+    });
+    return hits;
+  });
+}
+
+function _byteArrayToHex(arr) {
+  const u8arr = new Uint8Array(arr);
+  const hexs = [];
+  for (let i=0; i != u8arr.length; i += 1) {
```
Comment (radare) [inline src/agent/index.js]: spaces around =

File: src/agent/index.js [hunk 8]
```diff
@@ -1313,6 +1383,217 @@ function fridaVersion() {
...
+  throw new Error('Invalid hex string');
+}
+
+function _toPairs(hex) {
+  if ((hex.length % 2) !== 0) {
+    throw new Error('Odd-length string');
+  }
+
+  const pairs = [];
+  for (let i=0; i != hex.length; i += 2) {
```
Comment (radare) [inline src/agent/index.js]: spaces around =

File: src/agent/index.js [hunk 9]
```diff
@@ -1313,6 +1383,217 @@ function fridaVersion() {
...
+    pairs.push(hex.substr(i,2));
```
Comment (radare) [inline src/agent/index.js]: space after ,

**Comment by radare** [comment]
yep +1 for semistandard

File: src/io_frida.c [hunk 10]
```diff
@@ -623,6 +637,29 @@ static JsonObject *perform_request(RIOFrida *rf, JsonBuilder *builder, GBytes *d
...
 
+static void perform_request_unlocked(RIOFrida *rf, JsonBuilder *builder, GBytes *data, GBytes **bytes) {
+	JsonNode *root;
+	char *message;
+	GError *error = NULL;
+
+	json_builder_end_object (builder);
+	json_builder_end_object (builder);
+	root = json_builder_get_root (builder);
+	message = json_to_string (root, FALSE);
```
Comment (radare) [inline src/io_frida.c]: you can char * in here and JsonNode in there. no need to define variables at the beginning

**Pull request merged**